src/wasm.h

```diff
@@ -88,25 +88,37 @@ struct Name : public cashew::IString {
   }
 };
 
-// An index in a wasm module
-typedef uint32_t Index;
-
-// An address in linear memory. For now only wasm32
-struct Address {
-  typedef uint32_t address_t;
-  address_t addr;
-  Address() : addr(0) {}
-  Address(uint64_t a) : addr(static_cast<address_t>(a)) {
-    assert(a <= std::numeric_limits<address_t>::max());
-  }
-  Address& operator=(uint64_t a) {
-    assert(a <= std::numeric_limits<address_t>::max());
-    addr = static_cast<address_t>(a);
+// An index in a wasm module. This could be e.g. a local index or an index in
+// linear memory (i.e. an address).
+template <typename T>
+struct WasmIndex {
+  typedef T value_type;
+  T addr;
+  WasmIndex() : addr(0) {}
+  static void CheckValid(uint64_t v) {
+    // Allow -1 as an invalid sentinel value
+    assert(v <= std::numeric_limits<T>::max() ||
+	   v == static_cast<uint64_t>(-1));
+  }
+  WasmIndex(uint64_t a) : addr(static_cast<T>(a)) {
+    CheckValid(a);
+  }
+  WasmIndex& operator=(uint64_t a) {
+    CheckValid(a);
+    addr = static_cast<T>(a);
     return *this;
   }
-  operator address_t() const { return addr; }
-  Address& operator++() { ++addr; return *this; }
+  operator T() const { return addr; }
+  WasmIndex& operator++() { ++addr; return *this; }
+  WasmIndex operator++(int) {
```

Comment:
`operator++()` is pre-increment and `operator++(int)` is post-increment. The extra parameter is an unused dummy which just serves to give them different mangled names.